File: Synthesizer/Synthesizer/PlotDrawer.cpp
#include "PlotDrawer.h"
#include <iostream>
#include <string>
#include <fstream>
#include <thread>

void PlotDrawer::draw(
	const std::string & labelX,
	const std::string & labelY,
	const std::vector<double> & valuesX,
	const std::vector<double> & valuesY)
{
	std::cout << "Parsing params..." << std::endl;
	std::ofstream parameters("parameters.txt", std::ofstream::out);

	parameters << "regular" << "\n";
	parameters << labelX << "\n";
	parameters << labelY << "\n";
	parameters << vectorToString(valuesX) << "\n";
	parameters << vectorToString(valuesY) << "\n";

	parameters.close();

	std::cout << "Drawing plot..." << std::endl;
	system("python PlotDrawer.py &");

	std::cout << "Plot finished." << std::endl;
}

void PlotDrawer::draw(const std::vector<std::vector<double>>& matrix)
{
	std::cout << "Parsing params..." << std::endl;
	std::ofstream parameters("parameters.txt", std::ofstream::out);

	parameters << "grid" << "\n";
	for (const auto& row : matrix)
		parameters << vectorToString(row) << "\n";

	parameters.close();

	std::cout << "Drawing plot..." << std::endl;
	system("python PlotDrawer.py");

	std::cout << "Plot finished." << std::endl;
}

template<class T>
std::string PlotDrawer::vectorToString(const std::vector<T>& vec)
{
	std::string result{ "" };

	for (uint32_t i = 0; i < vec.size(); i++)
	{
		result += std::to_string(vec[i]);
		result += (i != vec.size() - 1 ? "," : "");
	}
	return result;
}

File: Synthesizer/Synthesizer/PlotDrawer.py
# -*- coding: utf-8 -*-

import matplotlib as mp
import numpy as np
import matplotlib.pyplot as plt

def regular(lines):
	labelX = lines[1]
	labelY = lines[2]
	valuesX = np.array(lines[3].split(',')).astype(np.float)
	valuesY = np.array(lines[4].split(',')).astype(np.float)

	plt.xlabel(labelX)
	plt.ylabel(labelY)
	plt.grid(color='black', linestyle='-', linewidth=0.5)

	plt.plot(valuesX, valuesY)

	plt.show()
	
def grid(lines):

	matrix = []
	for i in range(1, len(lines)):
		row = np.array(lines[i].split(',')).astype(np.float)
		matrix.append(row)

	plt.pcolor(matrix, cmap='Greys')
	plt.show()
		
parameters = open("parameters.txt")

input = parameters.readlines()

if input[0].rstrip() == "grid":
	grid(input)
if input[0].rstrip() == "regular":
	regular(input)